internal/cmd/agent/start.go

```diff
@@ -42,59 +44,62 @@ func start(ctx context.Context, kubeConfig, namespace, agentScope string, opts *
 		return err
 	}
 
-	agentInfo, err := register.Register(ctx, namespace, opts.ClusterID, kc)
+	// try to claim leadership lease without rate limiting
+	localConfig := rest.CopyConfig(kc)
+	localConfig.RateLimiter = ratelimit.None
+	k8s, err := kubernetes.NewForConfig(localConfig)
 	if err != nil {
 		return err
 	}
 
-	fleetNamespace, _, err := agentInfo.ClientConfig.Namespace()
-	if err != nil {
-		return err
-	}
+	leader.RunOrDie(ctx, namespace, "fleet-agent-lock", k8s, func(ctx context.Context) {
```

Comment:
why do we want to do the controllers registration inside the leader block? Controllers will not start receiving events until the factories are started in https://github.com/rancher/fleet/blob/master/internal/cmd/agent/controllers/controllers.go#L129 

I'd prefer to leave everything ready for the slave agent, so it takes less time to start handling events when it takes the leadership. And if there is any error when registering we know it before it is leader. Am I missing any advantage of waiting for it to be leader?

Reply:
I updated the explanation in the PR :)

Reply:
thanks!